dotCMS/src/main/java/com/dotcms/security/secret/ServiceIntegrationAPIImpl.java

```diff
@@ -0,0 +1,554 @@
+                        Collectors.toMap(serviceDescriptorMeta -> serviceDescriptorMeta
+                                        .getServiceDescriptor().getKey(), Function.identity(),
+                                (serviceDescriptor, serviceDescriptor2) -> serviceDescriptor));
+
+                CacheLocator.getCacheAdministrator()
+                        .put(DESCRIPTORS_MAPPED_BY_SERVICE_KEY, descriptorsByServiceKey,
+                                DESCRIPTORS_CACHE_GROUP);
+            }
+        }
+        return descriptorsByServiceKey;
+    }
+
+    @Override
+    public Optional<ServiceDescriptor> getServiceDescriptor(final String serviceKey,
+            final User user)
+            throws DotDataException, DotSecurityException {
+        final String serviceKeyLC = serviceKey.toLowerCase();
+        final ServiceDescriptorMeta serviceDescriptorMeta = getServiceDescriptorMap(user)
+                .get(serviceKeyLC);
+        return null == serviceDescriptorMeta ? Optional.empty()
+                : Optional.of(serviceDescriptorMeta.getServiceDescriptor());
+    }
+
+    @Override
+    public void createServiceDescriptor(final InputStream inputStream,
+            final User user) throws IOException, DotDataException, DotSecurityException {
+        if (userDoesNotHaveAccess(user)) {
+            throw new DotSecurityException(String.format(
+                    "Invalid attempt to create a service descriptors performed by user with id `%s`.",
+                    user.getUserId()));
+        }
+
+        final String ymlFilesPath = getServiceDescriptorDirectory();
+        final File basePath = new File(ymlFilesPath);
+        if (!basePath.exists()) {
+            basePath.mkdir();
+        }
+        Logger.debug(ServiceIntegrationAPIImpl.class,
+                () -> " ymlFiles are set under:  " + ymlFilesPath);
+
+        // Now validate the incoming file.. see if we're rewriting an existing file or attempting to re-use an already in use service-key.
+        final ServiceDescriptor serviceDescriptor = ymlMapper
+                .readValue(inputStream, ServiceDescriptor.class);
+
+        if (validateServiceDescriptor(serviceDescriptor)
+                && validateServiceDescriptorUniqueName(serviceDescriptor, user)) {
+
+            final String serviceKey = serviceDescriptor.getKey();
+            final File incomingFile = new File(basePath, String.format("%s.yml", serviceKey));
+            if (incomingFile.exists()) {
+                throw new DotDataException(
+                        String.format("Invalid attempt to override an existing file named '%s'.",
+                                incomingFile.getName()));
+            }
+
+            ymlMapper.writeValue(incomingFile, serviceDescriptor);
+
+            invalidateCache();
+        }
+    }
+
+    @Override
+    public void removeServiceIntegration(final String serviceKey, final User user,
+            boolean removeDescriptor)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Parameter 'removeDescriptor' is not assigned and could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4623512)